src/engine/readaheadmanager.cpp

```diff
@@ -23,88 +25,119 @@ ReadAheadManager::ReadAheadManager(CachingReader* pReader,
                                    LoopingControl* pLoopingControl)
         : m_pLoopingControl(pLoopingControl),
           m_pRateControl(NULL),
-          m_iCurrentPosition(0),
+          m_currentPosition(0),
           m_pReader(pReader),
           m_pCrossFadeBuffer(SampleUtil::alloc(MAX_BUFFER_LEN)) {
     DEBUG_ASSERT(m_pLoopingControl != NULL);
     DEBUG_ASSERT(m_pReader != NULL);
-    SampleUtil::clear(m_pCrossFadeBuffer, MAX_BUFFER_LEN);
 }
 
 ReadAheadManager::~ReadAheadManager() {
     SampleUtil::free(m_pCrossFadeBuffer);
 }
 
-SINT ReadAheadManager::getNextSamples(double dRate, CSAMPLE* buffer,
+SINT ReadAheadManager::getNextSamples(double dRate, CSAMPLE* pOutput,
         SINT requested_samples) {
     // TODO(XXX): Remove implicit assumption of 2 channels
     if (!even(requested_samples)) {
         qDebug() << "ERROR: Non-even requested_samples to ReadAheadManager::getNextSamples";
         requested_samples--;
     }
     bool in_reverse = dRate < 0;
-    SINT start_sample = m_iCurrentPosition;
+
     //qDebug() << "start" << start_sample << requested_samples;
-    SINT samples_needed = requested_samples;
-    CSAMPLE* base_buffer = buffer;
 
-    // A loop will only limit the amount we can read in one shot.
 
+    // A loop will only limit the amount we can read in one shot.
     const double loop_trigger = m_pLoopingControl->nextTrigger(
-            dRate, m_iCurrentPosition, 0, 0);
+            dRate, m_currentPosition, 0, 0);
     bool loop_active = loop_trigger != kNoTrigger;
     SINT preloop_samples = 0;
+    double samplesToLoopTrigger = 0.0;
 
+    SINT samples_from_reader = requested_samples;
     if (loop_active) {
-        SINT samples_available = in_reverse ?
-                m_iCurrentPosition - loop_trigger :
-                loop_trigger - m_iCurrentPosition;
-        if (samples_available < 0) {
-            samples_needed = 0;
+        samplesToLoopTrigger = in_reverse ?
+                m_currentPosition - loop_trigger :
+                loop_trigger - m_currentPosition;
+        if (samplesToLoopTrigger < 0) {
+            // We have already passed the loop trigger
+            samples_from_reader = 0;
         } else {
-            preloop_samples = samples_available;
-            samples_needed = math_clamp<SINT>(samples_needed, 0, samples_available);
+            // We can only read whole frames from the reader.
+            // Use ceil here, to be sure to reach the loop trigger.
+            preloop_samples = SampleUtil::ceilPlayPosToFrameStart(samplesToLoopTrigger,
+                    kNumChannels);
+            // clamp requested samples from the caller to the loop trigger point
+            samples_from_reader = math_clamp(requested_samples,
+                    static_cast<SINT>(0), preloop_samples);
         }
     }
 
+    int start_sample;
     if (in_reverse) {
-        start_sample = m_iCurrentPosition - samples_needed;
+        start_sample = SampleUtil::roundPlayPosToFrameStart(
```

Comment:
...and CachingReader already delivers the result samples in reverse order, so the "start" position should match this direction -> Inconsistent interface design of CachingReader.